### PR TITLE
[ASTGen] Implement Patterns

### PR DESCRIPTION
**Pull request by rintaro**
Expect `MissingPatternSyntax`.

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by ahoppen — APPROVED** [review]

File: lib/ASTGen/Sources/ASTGen/Patterns.swift [hunk 1]
```diff
@@ -45,6 +50,60 @@ extension ASTGenVisitor {
...
+      )
+    )
+  }
+
+  func generate(missingPattern node: MissingPatternSyntax) -> BridgedPattern {
+    fatalError("unimplemented")
+  }
+
+  func generate(tuplePattern node: TuplePatternSyntax) -> BridgedPattern {
+    if node.elements.count == 1 && node.elements.first!.label == nil {
```
Comment (bnbarham) [inline lib/ASTGen/Sources/ASTGen/Patterns.swift:70]:
```suggestion
    if node.elements.count == 1, let firstElement = node.elements.first, firstElement.label == nil {
```

And then `firstElement` can be re-used further down (`firstElement.pattern`).

File: include/swift/AST/ASTBridging.h [hunk 2]
```diff
@@ -1111,12 +1111,54 @@ void BridgedTypeRepr_dump(BridgedTypeRepr type);
...
+SWIFT_NAME(
+    "BridgedParenPattern.createParsed(_:lParenLoc:subPattern:rParenLoc:)")
+BridgedParenPattern BridgedParenPattern_createParsed(
+    BridgedASTContext cContext, BridgedSourceLoc cLParenLoc,
+    BridgedPattern cSubPattern, BridgedSourceLoc cRParenLoc);
+
+struct BridgedTuplePatternElt {
+  BridgedIdentifier Label;
+  BridgedSourceLoc LabelLoc;
+  BridgedPattern ThePattern;
```
Comment (bnbarham) [inline include/swift/AST/ASTBridging.h:1149]: Lowercase?

File: lib/ASTGen/Sources/ASTGen/Patterns.swift [hunk 3]
```diff
@@ -45,6 +50,60 @@ extension ASTGenVisitor {
...
   }
+
+  func generate(isTypePattern node: IsTypePatternSyntax) -> BridgedIsPattern {
+    return .createParsed(
+      self.ctx,
+      isLoc: self.generateSourceLoc(node.isKeyword),
+      typeExpr: .createParsed(
+        self.ctx,
+        type: self.generate(type: node.type)
+      )
```
Comment (hamishknight) [inline lib/ASTGen/Sources/ASTGen/Patterns.swift:61]: Don't we have a `generate` function for `TypeExpr`?

**Review by hamishknight — APPROVED** [review]

File: lib/ASTGen/Sources/ASTGen/Patterns.swift (hunk 3 above)
Comment (rintaro) [reply lib/ASTGen/Sources/ASTGen/Patterns.swift:61]: We do, but that only accept `TypeExprSyntax` instead of `TypeSyntax`.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

**Pull request merged**